### ISSUE TITLE
access to the records in each bin

### ISSUE THREAD

**Issue by gordonwoodhull** [crossfilter/crossfilter#25]
It seems in line with our general philosophy on this fork to provide access to the underlying row data for each bin in a group. 

This is useful for "complex reductions" such as determining the min, max, median of the rows. Currently one has to maintain an array which duplicates information already maintained by crossfilter groups in the [`groupIndex`](https://github.com/crossfilter/crossfilter/blob/master/src/crossfilter.js#L371) variable.

There are some annoying design problems:
 * should the method return the original records (`allRecords`), or indices to the records (`allIndices`)?
 * is there any faster way to do this than to iterate over `groupIndex` and generate a whole bunch of arrays? It's just O(N) but it's a lot of memory allocation.

But I don't think those are as annoying or as error-prone as maintaining duplicate information as a crossfilter client.

I looked for an existing issue here or on square/crossfilter, but couldn't find one, so I'm opening this. I'm sure it's been mentioned many times before, so let's add links to prior discussions/applications as we find them.

**Comment by alan-unravel** [comment]
Could we make it a generator? seems like that might work for reductions without The Space Complexity, the user can just exhaust the iterator if they wish to have a carbon copy.

**Comment by gordonwoodhull** [comment]
So instead of returning all the records, `allRecords` could return `{key, recordsFn}` - yeah! That makes sense.

**Comment by gordonwoodhull** [comment]
Or I guess you mean an ES6 generator - I'm not familiar with those but it's probably syntactic sugar on the same idea.

**Comment by alan-unravel** [comment]
I did but I'm sure they are quite similar

**Comment by esjewett** [comment]
Great idea. I agree that we should provide a function (or generator if/when we go ES6) that will return all the records so that we don't generate the arrays of records if it's not necessary.

Question: Instead of providing an array of functions to get at records in each bin, should we provide the functions in the reducer? If the reducers want to expose the records, they can attach the function to the returned group object. Or they can call the function internally to do calculations based on the current record set.

**Comment by esjewett** [comment]
Oh, I'd also say that I definitely think we should return records and not indices. I don't think we want to guarantee that the order of records we return from any of these functions match any internal ordering that we maintain.

**Comment by gordonwoodhull** [comment]
Well, it's not an internal ordering, it's the indices of the records in the original data passed to crossfilte. But I'm okay with either interface.

That's a smart idea to provide access to the reduce functions rather than external access. That will simplify the interface and is the more common use case. I guess we could also add a convenience method like `reduceSum` e.g. `reduceRows` so that the reduction is simply returning the rows. Then min/max/median become one-liners.

**Comment by esjewett** [comment]
Yeah, true. I hadn't remembered that - though we don't really guarantee that that is the case, do we (that the indices match the passed-in data)? I'm also not sure what happens once you start using crossfilter.add and crossfilter.remove. Maybe the function could return an array of records by default, but could take a boolean parameter and if true it would return an array of indices? There are definitely situations in which the array of indices will be much more efficient, even if it gets dicey in more complex scenarios. But I think the default behavior should be to return records. Returning indices should probably come with a big warning sign if we expose that at all.

**Comment by gordonwoodhull** [comment]
I am pretty sure that crossfilter just keeps the original array. Then add and remove change the indices in the way you'd expect. You're right this is somewhat dangerous and might close doors to later optimizations or features.

I don't think if it would be any less efficient to return the records, since everything in JavaScript is by-reference so nothing is copied. You get back references to exactly the same objects that you passed in, so it's like an array of pointers instead of an array of integers, which are usually the same efficiency. (Er, maybe JavaScript integers are 32-bit instead of 64-bit pointers, but who cares.)

My thought was that the indices are somewhat more powerful, since they refer to the original data rather than plunking you into the data without knowing its context. But I'm not sure this is actually useful. I'm now convinced we should just return the rows/records.

**Comment by tannerlinsley** [comment]
Where are we on this thought train?  It seems like this could in face enable some very power features having to do with (but not limited to) post-processing aggregations?

**Comment by gordonwoodhull** [comment]
I like it and I think it's pretty clear cut except for the name. "reduceRows" is wrong because it's really  "don't reduce".

I think it should be possible to implement without performance degradation to other features, even if it itself is potentially expensive. "Pay for what you use."

**Comment by derekperkins** [comment]
:+1:
